File: ex167.go
func binarySearch(nums []int, target int, lo int) int {
    hi := len(nums)-1
    for lo <= hi {
        mid := lo + (hi - lo) / 2
        if nums[mid] < target {
            lo = mid+1
        } else if nums[mid] > target {
            hi = mid-1
        } else {
            return mid
        }
    }
    return -1
}

func twoSum(nums []int, target int) []int {
    
    for i := 0; i < len(nums); i++ {
        x := binarySearch(nums, target - nums[i], i+1)
        if x >= 0 {
            return []int{i+1, x+1}
        }
    }
    return nil
}

File: leetcode/ex167.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        vector<int> pair;
        for (int i = 0; i < numbers.size(); i++) {
            int v = target - numbers[i];
            int lo = i+1, hi = numbers.size()-1;
            while (lo <= hi) {
                int mid = (lo + hi) / 2;
                if (numbers[mid] < v)
                    lo = mid+1;
                else if (numbers[mid] > v)
                    hi = mid-1;
                else {
                    pair.push_back(i+1);
                    pair.push_back(mid+1);
                    return pair;
                }
            }
        }
        return pair;
    }
};
